service/src/java/org/apache/hive/service/servlet/OTELExporter.java

```diff
@@ -0,0 +1,225 @@
+        for (QueryDisplay.TaskDisplay task : lQuery.getQueryDisplay().getTaskDisplays()) {
+          if (task.getReturnValue() != null && task.getEndTime() != null && !queryIdToTasksMap.get(queryID).contains(task.getTaskId())) {
+            queryIdToTasksMap.get(queryID).add(task.getTaskId());
+            Context parentContext = Context.current().with(rootspan);
+            tracer.spanBuilder(queryID+ " - " + task.getTaskId() + " - live")
+                    .setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                    .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan()
+                    .end(task.getEndTime(), TimeUnit.MILLISECONDS);
+          }
+        }
+      } else {
+        // In case of live queries being seen for first time and has initialized its queryDisplay
+        rootspan = tracer.spanBuilder(queryID + " - live")
+                .startSpan();
+        List<String> completedTasks = new ArrayList<>();
+        Context parentContext = Context.current().with(rootspan);
+
+        Span initSpan = tracer.spanBuilder(queryID + " - live").setParent(parentContext).startSpan()
+                .setAttribute("queryID", lQuery.getQueryDisplay().getQueryId())
+                .setAttribute("queryString", lQuery.getQueryDisplay().getQueryString())
+                .setAttribute("Begin Time", lQuery.getBeginTime());
+        if (lQuery.getQueryDisplay().getErrorMessage() != null) {
+          initSpan.setAttribute("Error Message", lQuery.getQueryDisplay().getErrorMessage());
+        }
+        initSpan.end();
+
+        for (QueryDisplay.TaskDisplay task : lQuery.getQueryDisplay().getTaskDisplays()) {
+          if (task.getReturnValue() != null && task.getEndTime() != null) {
+            completedTasks.add(task.getTaskId());
+            parentContext = Context.current().with(rootspan);
+            tracer.spanBuilder(queryID + " - " + task.getTaskId() + " - live")
+                    .setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                    .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan()
+                    .end(task.getEndTime(), TimeUnit.MILLISECONDS);
+          }
+        }
+        queryIdToSpanMap.put(queryID,rootspan);
+        queryIdToTasksMap.put(queryID,completedTasks);
+      }
+    }
+
+    List<String> historicalQueryIDs = new ArrayList<>();
+    for (QueryInfo hQuery : historicalQueries) {
+      String hQueryId = hQuery.getQueryDisplay().getQueryId();
+      historicalQueryIDs.add(hQueryId);
+      Span rootspan = queryIdToSpanMap.get(hQueryId);
+
+      //For queries that were live till last loop but have ended before start of this loop
+      if (rootspan != null) {
+        for (QueryDisplay.TaskDisplay task : hQuery.getQueryDisplay().getTaskDisplays()) {
+          if (!queryIdToTasksMap.get(hQueryId).contains(task.getTaskId())) {
+            queryIdToTasksMap.get(hQueryId).add(task.getTaskId());
```

Comment:
can we do ``queryIdToTasksMap.remove(hQueryId)`` once & store it in a variable and reuse & why do we need to add here, we won't ever come back for this query again, right?

Reply:
Yes, changed the Map to store the tasks in a set and have initially stored the set in a variable by removing it from `queryIdToTasksMap`. Also, have removed the add of new taskID as its not required.